src/gun_pool_events_h.erl

```diff
@@ -0,0 +1,157 @@
+%% Copyright (c) 2021, Loïc Hoguin <essen@ninenines.eu>
+%%
+%% Permission to use, copy, modify, and/or distribute this software for any
+%% purpose with or without fee is hereby granted, provided that the above
+%% copyright notice and this permission notice appear in all copies.
+%%
+%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+%% ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+%% OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+-module(gun_pool_events_h).
+
+-export([init/2]).
+-export([domain_lookup_start/2]).
+-export([domain_lookup_end/2]).
+-export([connect_start/2]).
+-export([connect_end/2]).
+-export([tls_handshake_start/2]).
+-export([tls_handshake_end/2]).
+-export([request_start/2]).
+-export([request_headers/2]).
+-export([request_end/2]).
+-export([push_promise_start/2]).
+-export([push_promise_end/2]).
+-export([response_start/2]).
+-export([response_inform/2]).
+-export([response_headers/2]).
+-export([response_trailers/2]).
+-export([response_end/2]).
+-export([ws_upgrade/2]).
+-export([ws_recv_frame_start/2]).
+-export([ws_recv_frame_header/2]).
+-export([ws_recv_frame_end/2]).
+-export([ws_send_frame_start/2]).
+-export([ws_send_frame_end/2]).
+-export([protocol_changed/2]).
+-export([origin_changed/2]).
+-export([cancel/2]).
+-export([disconnect/2]).
+-export([terminate/2]).
+
+init(Event, State) ->
+	propagate(Event, State, ?FUNCTION_NAME).
+
+domain_lookup_start(Event, State) ->
+	propagate(Event, State, ?FUNCTION_NAME).
+
+domain_lookup_end(Event, State) ->
+	propagate(Event, State, ?FUNCTION_NAME).
+
+connect_start(Event, State) ->
+	propagate(Event, State, ?FUNCTION_NAME).
+
+connect_end(Event, State) ->
+	propagate(Event, State, ?FUNCTION_NAME).
+
+tls_handshake_start(Event, State) ->
+	propagate(Event, State, ?FUNCTION_NAME).
+
+tls_handshake_end(Event, State) ->
+	propagate(Event, State, ?FUNCTION_NAME).
+
+request_start(Event=#{stream_ref := StreamRef}, State=#{table := Tid}) ->
+	_ = ets:update_counter(Tid, self(), +1, {self(), 0}),
```

Comment:
This module will +1 or -1 the current stream count based on events. This is why the pool is not strict at the moment, this occurs after the pid has been given to the calling process.